### ISSUE TITLE
Function.apply should handle callable objects for --strong

### ISSUE THREAD

**Issue by johnniwinther** [dart-lang/sdk#32375]
Dart2js implementation of `Function.apply` doesn't handle callable object for `--strong` where such classes no longer implement Function:

```
class C { call() => 42; }
method(o) => Function.apply(o, []);
main() => method(new C());
```

`corelib_2/apply_test/01` fails because of this.

**Comment by lrhn** [comment]
The problem is indeed in the test. The test should be changed to test two things:

- That the above test fails because `o` is not assignable to `Function`,
- and if `o` has type `C`, then the tear-off gets passed and the call works.

**Comment by eernstg** [comment]
The semantics of function invocation `f(args)` is currently specified to involve (1) some notion of invocation of a function object on `args` (this is allowed to differ from invoking an instance method named `call` on `args`), and (2) if we're invoking some non-function object then `f(args)` is executed as the instance method invocation `f.call(args)`.

This ensures that dynamic invocation will apply both the (possibly primitive) function object invocation and the `.call` instance method invocation (such that callable objects will still work in the dynamic case).

But we have no special exceptions for statically checked calls: If a statically checked call site ends up making an attempt to invoke a callable object, it must be able to invoke the `.call` method rather than executing a primitive function invocation (if there is any difference, I'm just saying that it's allowed to make those two things different).

It just happens to be true that no callable object can have a subtype of any function type, so, by heap and expression soundness, invocations of a function with a function type can go ahead an use primitive function invocation.

For invocations of an instance of type `Function`, we may be able to establish certain invariants (like: "it could be a callable object or a function object" or (more likely, in the near future): "it is definitely a function object"), and then the semantics should correspond to that: If it can only be a function object then primitive invocation can be used and the callable object will be an error; if it can be both then both modes of invocation must be attempted.

For consistency, `Function.apply` should be as capable as the dynamic invocation (and, in principle, typed invocations are equally capable), so if only function objects can have type `Function`, `Function.apply` should _stop_ requiring its first argument to be a `Function`.

In other words, it is a reasonable expectation that we will have the following:
```dart
class C { call(x) => x; }
main() {
  dynamic d = new C();
  print(d(42)); // Works, prints '42'.
  Function.apply(d, [42]); // Should work and print '42',
  // .. simply because the same as a dynamic function invocation _does_ work.
}
```

It can be construed to be a breaking change to make it `Function.apply(Object d, ...)`, but we don't usually consider missing failures to be breakage, so we should be able to make that change.
